File: platforms/macos/tests/FrequencyAdjustmentPreferenceTests.cpp
#include "../FrequencyAdjustmentPreference.h"
#include <cassert>
#include <cstring>

int main()
{
    using namespace metasequoia::mac;
    assert(NormalizeFrequencyAdjustmentCount(1) == 1);
    assert(NormalizeFrequencyAdjustmentCount(6) == 6);
    assert(NormalizeFrequencyAdjustmentCount(0) == 1);
    assert(NormalizeFrequencyAdjustmentCount(7) == 1);
    assert(std::strcmp(NormalizeFrequencyAdjustmentMode(nullptr), "promote") == 0);
    assert(std::strcmp(NormalizeFrequencyAdjustmentMode("unknown"), "promote") == 0);
    assert(std::strcmp(NormalizeFrequencyAdjustmentMode("pin"), "pin") == 0);
    assert(FrequencyAdjustmentModeOptionIndex("pin") == 0);
    assert(FrequencyAdjustmentModeOptionIndex("halve") == 1);
    assert(FrequencyAdjustmentModeOptionIndex("linear") == 2);
    assert(FrequencyAdjustmentModeOptionIndex("promote") == 3);
    assert(std::strcmp(FrequencyAdjustmentModeForOptionIndex(0), "pin") == 0);
    assert(std::strcmp(FrequencyAdjustmentModeForOptionIndex(99), "promote") == 0);
    assert(EngineFrequencyOptions(false, "pin", 3, 4).mode == FrequencyAdjustmentMode::Disabled);
    const auto options = EngineFrequencyOptions(true, "linear", 3, 4);
    assert(options.mode == FrequencyAdjustmentMode::Linear);
    assert(options.trigger_count == 3 && options.linear_step == 4);
    return 0;
}

File: platforms/ios/tests/settings/FrequencyAdjustmentPreferenceTests.swift
import Foundation

@main
enum FrequencyAdjustmentPreferenceTests {
  static func main() {
    precondition(FrequencyAdjustmentPreference.resolvedMode(nil) == .promote)
    precondition(FrequencyAdjustmentPreference.resolvedMode("pin") == .pin)
    precondition(FrequencyAdjustmentPreference.resolvedMode("halve") == .halve)
    precondition(FrequencyAdjustmentPreference.resolvedMode("linear") == .linear)
    precondition(FrequencyAdjustmentPreference.resolvedMode("promote") == .promote)
    precondition(FrequencyAdjustmentPreference.resolvedMode("disabled") == .disabled)
    precondition(FrequencyAdjustmentPreference.resolvedMode("unknown") == .promote)
    precondition(FrequencyAdjustmentPreference.resolvedCount(nil) == 1)
    precondition(FrequencyAdjustmentPreference.resolvedCount(3) == 3)
    precondition(FrequencyAdjustmentPreference.resolvedCount(1) == 1)
    precondition(FrequencyAdjustmentPreference.resolvedCount(6) == 6)
    precondition(FrequencyAdjustmentPreference.resolvedCount(0) == 1)
    precondition(FrequencyAdjustmentPreference.resolvedCount(10) == 10)
    precondition(FrequencyAdjustmentPreference.resolvedCount(11) == 1)
    precondition(FrequencyAdjustmentPreference.resolvedCount(NSNumber(value: 4)) == 4)
    precondition(FrequencyAdjustmentMode.pin.title == "一次置顶")
    precondition(FrequencyAdjustmentMode.promote.title == "一次置前")
    precondition(FrequencyAdjustmentMode.disabled.title == "不调频")
  }
}
